src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -925,301 +1296,410 @@ internal TransitionRegex<TSet> CreateDerivative()
+                    //if X is not nullable then XZ is maintaned as is, and YZ is pruned
+                    //e.g. (a{0,5}?|abc|b+)c* reduces to c*
+                    //---
+                    //in a concatenation XZ where X is not an alternation, both X and Z are pruned
+                    //e.g. a{0,5}?b{0,5}? reduces to ()
+                    prunedNode = _left._kind == SymbolicRegexNodeKind.OrderedOr ?
+                        (_left._left!.IsNullableFor(context) ?
+                            CreateConcat(_builder, _left._left, _right).PruneLowerPriorityThanNullability(context) :
+                            OrderedOr(_builder, CreateConcat(_builder, _left._left, _right), CreateConcat(_builder, _left._right!, _right).PruneLowerPriorityThanNullability(context))) :
+                        CreateConcat(_builder, _left.PruneLowerPriorityThanNullability(context), _right.PruneLowerPriorityThanNullability(context));
+                    break;
+
+                case SymbolicRegexNodeKind.Loop when _info.IsLazyLoop && _lower == 0:
+                    //lazy nullable loop reduces to (), i.e., the loop body is just forgotten
+                    prunedNode = _builder.Epsilon;
+                    break;
+
+                case SymbolicRegexNodeKind.Effect:
+                    //Effects are maintained and the pruning is propagated to the body of the effect
+                    Debug.Assert(_left is not null && _right is not null);
+                    prunedNode = CreateEffect(_builder, _left.PruneLowerPriorityThanNullability(context), _right);
+                    break;
+
+                default:
+                    //In all other remaining cases no pruning takes place
+                    prunedNode = this;
+                    break;
             }
 
+            _builder._pruneCache[key] = prunedNode;
+            return prunedNode;
+        }
+
+        /// <summary>
+        /// Takes the derivative of the symbolic regex for the given element, which must be either
+        /// a minterm (i.e. a class of characters that have identical behavior for all sets in the pattern)
+        /// or a singleton set, and a context, which encodes the relevant information about the surrounding
+        /// characters for deciding the nullability of anchors.
+        /// </summary>
+        /// <remarks>
+        /// This derivative differs from ones familiar from literature in several ways:
+        /// -Ordered alternations are properly handled and any choices represented as alternations are ordered in a way
+        ///  that matches the backtracking engines.
+        /// -Some nodes, namely <see cref="SymbolicRegexNodeKind.CaptureStart"/> and <see cref="SymbolicRegexNodeKind.CaptureEnd"/>,
+        ///  will produce <see cref="SymbolicRegexNodeKind.Effect"/> nodes, which indicate effects to be applied on
+        ///  transitions using this derivative. The derivative must be further post-processed with a function that strips
```

Comment:
Can you add some examples of what it means to apply effects on transitions and also interpret effects?

Reply:
Done!